src/platforms/common/module.cpp

```diff
@@ -46,6 +46,169 @@
 
 pthread_mutex_t px4_modules_mutex = PTHREAD_MUTEX_INITIALIZER;
 
+BlockingList<ModuleBaseInterface *> _px4_modules_list;
+
+ModuleBaseInterface *get_module_instance(const char *name)
+{
+	auto lg = _px4_modules_list.getLockGuard();
+
+	// search list
+	for (ModuleBaseInterface *module : _px4_modules_list) {
+		const bool name_match = (strcmp(module->get_name(), name) == 0);
+
+		if (name_match) {
+			return module;
+		}
+	}
+
+	return nullptr;
+}
+
+bool module_running(const char *name)
+{
+	// search list
+	ModuleBaseInterface *module = get_module_instance(name);
+
+	if (module != nullptr) {
+		return module->running();
+	}
+
+	return false;
+}
+
+/**
+ * @brief Waits until object is initialized, (from the new thread). This can be called from task_spawn().
+ * @return Returns 0 iff successful, -1 on timeout or otherwise.
+ */
+int module_wait_until_running(const char *name)
+{
+	int i = 0;
+
+	ModuleBaseInterface *object = nullptr;
+
+	do {
+		object = get_module_instance(name);
+
+		// Wait up to 1 s
+		px4_usleep(2500);
+
+	} while (!object && ++i < 400);
+
+	if (i == 400) {
+		PX4_ERR("Timed out while waiting for thread to start");
+		return -1;
+	}
+
+	return 0;
+}
+
+int module_stop(const char *name)
+{
+	int ret = 0;
+	ModuleBaseInterface::lock_module();
+
+	if (module_running(name)) {
+
+		ModuleBaseInterface *object = nullptr;
+		unsigned int i = 0;
+
+		do {
+			// search for module again to request stop
+			object = get_module_instance(name);
+
+			if (object != nullptr) {
```

Comment:
This is a bit confusing because there are checks for `object != nullptr` in several places.
Could you invert the logic to:

```
object = get_module_instance(name);
if (object == nullptr) {
    break;
    ...
```

And with a scoped lock you could even get rid of the `do` `while` and just `return` directly.

Reply:
The locking and structure is a bit weird because it's expecting the task to kill itself cleanly, so it's sleeping, checking, etc in between.

I'll review it again.

Reply:
I still think you could turn that around for better readability.